File: libzahl/src/zlsh.c
/* See LICENSE file for copyright and license details. */
#include "internals.h"


void
zlsh(z_t a, z_t b, size_t bits)
{
	size_t i, chars, cbits;
	zahl_char_t carry = 0, tcarry;

	if (unlikely(zzero(b))) {
		SET_SIGNUM(a, 0);
		return;
	}

	chars = FLOOR_BITS_TO_CHARS(bits);
	bits = BITS_IN_LAST_CHAR(bits);
	cbits = BITS_PER_CHAR - bits;

	ENSURE_SIZE(a, b->used + chars + 1);
	if (likely(a == b)) {
		zmemmoveb(a->chars + chars, b->chars, b->used);
	} else {
		zmemcpy(a->chars + chars, b->chars, b->used);
	}
	zmemset_precise(a->chars, 0, chars);
	a->used = b->used + chars;

	if (likely(bits)) { /* This if statement is very important in C. */
		for (i = chars; i < a->used; i++) {
			tcarry = a->chars[i] >> cbits;
			a->chars[i] <<= bits;
			a->chars[i] |= carry;
			carry = tcarry;
		}
		if (carry)
			a->chars[a->used++] = carry;
	}

	SET_SIGNUM(a, zsignum(b));
}

File: project/llm_from_c_output/libzahl/zlsh.rs
/* See LICENSE file for copyright and license details. */

use std::cmp::Ordering;
use std::mem;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Z {
    chars: Vec<u8>,
    used: usize,
    signum: i8,
}

impl Z {
    fn zero() -> Self {
        Z {
            chars: Vec::new(),
            used: 0,
            signum: 0,
        }
    }

    fn is_zero(&self) -> bool {
        self.signum == 0
    }

    fn signum(&self) -> i8 {
        self.signum
    }

    fn ensure_size(&mut self, size: usize) {
        if self.chars.len() < size {
            self.chars.resize(size, 0);
        }
    }

    fn zmemmoveb(&mut self, dest: usize, src: usize, len: usize) {
        if dest == src || len == 0 {
            return;
        }
        if dest < src {
            for i in 0..len {
                self.chars[dest + i] = self.chars[src + i];
            }
        } else {
            for i in (0..len).rev() {
                self.chars[dest + i] = self.chars[src + i];
            }
        }
    }
}

const BITS_PER_CHAR: usize = 8;

fn floor_bits_to_chars(bits: usize) -> usize {
    bits / BITS_PER_CHAR
}

fn bits_in_last_char(bits: usize) -> usize {
    bits % BITS_PER_CHAR
}

pub fn zlsh(a: &mut Z, b: &Z, bits: usize) {
    if b.is_zero() {
        a.signum = 0;
        return;
    }

    let chars = floor_bits_to_chars(bits);
    let bits = bits_in_last_char(bits);
    let cbits = BITS_PER_CHAR - bits;

    a.ensure_size(b.used + chars + 1);
    if a as *const _ == b as *const _ {
        a.zmemmoveb(chars, 0, b.used);
    } else {
        a.chars[chars..chars + b.used].copy_from_slice(&b.chars[..b.used]);
    }
    a.chars[..chars].fill(0);
    a.used = b.used + chars;

    if bits != 0 {
        let mut carry: u8 = 0;
        for i in chars..a.used {
            let tcarry = a.chars[i] >> cbits;
            a.chars[i] <<= bits;
            a.chars[i] |= carry;
            carry = tcarry;
        }
        if carry != 0 {
            a.chars[a.used] = carry;
            a.used += 1;
        }
    }

    a.signum = b.signum();
}